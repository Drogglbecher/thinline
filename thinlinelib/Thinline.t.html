<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=struct.Thinline.html">
</head>
<body>
    <p>Redirecting to <a href="struct.Thinline.html">struct.Thinline.html</a>...</p>
    <script>location.replace("struct.Thinline.html" + location.search + location.hash);</script>
</body>
</html>