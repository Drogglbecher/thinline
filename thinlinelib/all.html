<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><meta name="keywords" content="rust, rustlang, rust-lang"><title>List of all items in this crate</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../dark.css"><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script src="../storage.js"></script></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!"><img src="../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices"></div></div><script src="../theme.js"></script><nav class="sub"><form class="search-form js-only"><div class="search-container"><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><a id="settings-menu" href="../settings.html"><img src="../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class='fqn'><span class='out-of-band'><span id='render-detail'><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class='inner'>&#x2212;</span>]</a></span>
     </span>
     <span class='in-band'>List of all items</span></h1><h3 id='Structs'>Structs</h3><ul class='structs docblock'><li><a href='struct.Thinline.html'>Thinline</a></li><li><a href='analysis/struct.Analysis.html'>analysis::Analysis</a></li><li><a href='analysis/struct.Argument.html'>analysis::Argument</a></li><li><a href='analysis/struct.Description.html'>analysis::Description</a></li><li><a href='analysis/struct.Enum.html'>analysis::Enum</a></li><li><a href='analysis/struct.Function.html'>analysis::Function</a></li><li><a href='analysis/struct.ProjectFile.html'>analysis::ProjectFile</a></li><li><a href='entity/struct.Entity.html'>entity::Entity</a></li><li><a href='language_type/struct.C.html'>language_type::C</a></li><li><a href='language_type/struct.Cpp.html'>language_type::Cpp</a></li><li><a href='language_type/struct.Python.html'>language_type::Python</a></li><li><a href='project_parameters/struct.BuildScript.html'>project_parameters::BuildScript</a></li><li><a href='project_parameters/struct.ProjectParameters.html'>project_parameters::ProjectParameters</a></li><li><a href='stubs/struct.Stub.html'>stubs::Stub</a></li><li><a href='stubs/struct.Stubs.html'>stubs::Stubs</a></li><li><a href='synthesis/struct.Synthesis.html'>synthesis::Synthesis</a></li><li><a href='synthesis/struct.TestClass.html'>synthesis::TestClass</a></li><li><a href='synthesis/struct.TestFile.html'>synthesis::TestFile</a></li><li><a href='synthesis/struct.TestFunction.html'>synthesis::TestFunction</a></li></ul><h3 id='Enums'>Enums</h3><ul class='enums docblock'><li><a href='entity/enum.EntityType.html'>entity::EntityType</a></li><li><a href='synthesis/enum.StubContextType.html'>synthesis::StubContextType</a></li></ul><h3 id='Traits'>Traits</h3><ul class='traits docblock'><li><a href='entity/trait.EntityConversion.html'>entity::EntityConversion</a></li><li><a href='language_type/trait.LanguageType.html'>language_type::LanguageType</a></li><li><a href='synthesis/trait.StubContextConversion.html'>synthesis::StubContextConversion</a></li><li><a href='value_parser/trait.ValueParser.html'>value_parser::ValueParser</a></li><li><a href='value_parser/trait.VecStringConversion.html'>value_parser::VecStringConversion</a></li></ul></section><section id="search" class="content hidden"></section><section class="footer"></section><aside id="help" class="hidden"><div><h1 class="hidden">Help</h1><div class="shortcuts"><h2>Keyboard Shortcuts</h2><dl><dt><kbd>?</kbd></dt><dd>Show this help dialog</dd><dt><kbd>S</kbd></dt><dd>Focus the search field</dd><dt><kbd>↑</kbd></dt><dd>Move up in search results</dd><dt><kbd>↓</kbd></dt><dd>Move down in search results</dd><dt><kbd>↹</kbd></dt><dd>Switch tab</dd><dt><kbd>&#9166;</kbd></dt><dd>Go to active search result</dd><dt><kbd>+</kbd></dt><dd>Expand all sections</dd><dt><kbd>-</kbd></dt><dd>Collapse all sections</dd></dl></div><div class="infos"><h2>Search Tricks</h2><p>Prefix searches with a type followed by a colon (e.g. <code>fn:</code>) to restrict the search to a given type.</p><p>Accepted types are: <code>fn</code>, <code>mod</code>, <code>struct</code>, <code>enum</code>, <code>trait</code>, <code>type</code>, <code>macro</code>, and <code>const</code>.</p><p>Search functions by type signature (e.g. <code>vec -> usize</code> or <code>* -> vec</code>)</p><p>Search multiple things at once by splitting your query with comma (e.g. <code>str,u8</code> or <code>String,struct:Vec,test</code>)</p></div></div></aside><script>window.rootPath = "../";window.currentCrate = "thinlinelib";</script><script src="../aliases.js"></script><script src="../main.js"></script><script defer src="../search-index.js"></script></body></html>