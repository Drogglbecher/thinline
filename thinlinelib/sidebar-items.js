initSidebarItems({"mod":[["analysis",""],["entity",""],["language_type",""],["project_parameters",""],["stubs",""],["synthesis",""],["value_parser",""]],"struct":[["Thinline","Global structure representing the `Thinline` lib."]]});