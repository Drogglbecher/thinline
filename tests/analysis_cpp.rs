extern crate snapshot;
extern crate thinlinelib;

#[cfg(test)]
mod cpp {
    use snapshot::snapshot;
    use std::path::Path;
    use thinlinelib::{analysis::Analysis, entity::EntityType, language_type::Cpp};

    #[test]
    fn cpp() {
        // test_collect_sources
        {
            // should_succeed when_directory_is_valid
            {
                // Given
                let analysis: Analysis<Cpp> = Analysis::new();

                // When
                let cpp_test_src_path = Path::new("tests").join("testdata").join("cpp_sources");
                assert!(
                    analysis
                        .collect_sources(&cpp_test_src_path, &[String::from(".")])
                        .is_ok()
                );

                // Then
                assert_eq!(analysis.project_files().len(), 2);
            }

            // should_fail
            {
                // when_directory_not_existing
                {
                    // Given
                    let analysis: Analysis<Cpp> = Analysis::new();

                    // When
                    let cpp_test_src_path = Path::new("not").join("existing");

                    // Then
                    assert!(
                        analysis
                            .collect_sources(&cpp_test_src_path, &[String::from(".")])
                            .is_err()
                    );
                }

                // when_path_is_no_directory
                {
                    // Given
                    let analysis: Analysis<Cpp> = Analysis::new();

                    // When
                    let cpp_test_src_path = Path::new("tests").join("lib.rs");

                    // Then
                    assert!(
                        analysis
                            .collect_sources(&cpp_test_src_path, &[String::from(".")])
                            .is_err()
                    );
                }
            }
        }
    }

    fn extract_entities_cpp() -> Vec<EntityType> {
        let analysis: Analysis<Cpp> = Analysis::new();
        let cpp_test_src_path = Path::new("tests").join("testdata").join("analysis");
        assert!(
            analysis
                .collect_sources(&cpp_test_src_path, &[String::from(".")])
                .is_ok()
        );

        assert!(analysis.extract_entities().is_ok());

        let project_files = analysis.project_files();

        assert_eq!(project_files.len(), 1);

        let entities = project_files[0].entities();
        assert_eq!(entities.len(), 1);
        entities[0].clone().entities
    }

    #[cfg(target_os = "linux")]
    #[snapshot]
    fn extract_entities_linux_cpp() -> Vec<EntityType> {
        extract_entities_cpp()
    }

    #[cfg(target_os = "windows")]
    #[snapshot]
    fn extract_entities_windows_cpp() -> Vec<EntityType> {
        extract_entities_cpp()
    }
}
