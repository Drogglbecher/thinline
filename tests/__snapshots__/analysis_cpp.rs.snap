{
  "analysis_cpp::cpp::extract_entities_linux_cpp": {
    "file": "tests/analysis_cpp.rs",
    "module_path": "analysis_cpp::cpp",
    "test_function": "extract_entities_linux_cpp",
    "recorded_value": [
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Entity": {
                "description": {
                  "lines": [
                    "#TESTCLASS(c1)",
                    "#SET_UP_CONTEXT:",
                    "this->class_inst = new c1();",
                    "#TEAR_DOWN_CONTEXT:",
                    "delete this->class_inst;",
                    "this->class_inst = nullptr;",
                    "#CLASS_CONTEXT:",
                    "c1 * class_inst;"
                  ]
                },
                "entities": [
                  {
                    "Function": {
                      "arguments": [
                        {
                          "atype": "unsigned int",
                          "name": "no1",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no2",
                          "value": null
                        }
                      ],
                      "description": {
                        "lines": [
                          "#TESTCASE(c1::AddTwoNumbers)",
                          "unsigned int no1 = 5;",
                          "#EQ[this->class_inst->TL_FCT(no1:no1,no2:10)=>15]",
                          "#LT[this->class_inst->TL_FCT(no1:no1,no2:10)=>30]"
                        ]
                      },
                      "name": "add_two_numbers",
                      "return_type": "unsigned int"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c1",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c1",
                      "return_type": "void"
                    }
                  }
                ],
                "name": "c1"
              }
            },
            {
              "Entity": {
                "description": {
                  "lines": [
                    "#TESTCLASS(c2)",
                    "#SET_UP_CONTEXT:",
                    "this->class_inst = new c2();",
                    "#TEAR_DOWN_CONTEXT:",
                    "delete this->class_inst;",
                    "this->class_inst = nullptr;",
                    "#CLASS_CONTEXT:",
                    "c2 * class_inst;"
                  ]
                },
                "entities": [
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c2",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c2",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [
                        {
                          "atype": "unsigned int",
                          "name": "no1",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no2",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no3",
                          "value": null
                        }
                      ],
                      "description": {
                        "lines": [
                          "#TESTCASE(c2::AddThreeNumbers)",
                          "unsigned int no1 = 5;",
                          "unsigned int no2 = 10;",
                          "#EQ[this->class_inst->TL_FCT(no1:no1,no2:no2,no3:5)=>20]"
                        ]
                      },
                      "name": "add_three_numbers",
                      "return_type": "unsigned int"
                    }
                  }
                ],
                "name": "c2"
              }
            }
          ],
          "name": "ns1"
        }
      },
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Entity": {
                "description": null,
                "entities": [
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c3",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c3",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "return5",
                      "return_type": "unsigned int"
                    }
                  }
                ],
                "name": "c3"
              }
            }
          ],
          "name": "ns2"
        }
      }
    ]
  },
  "analysis_cpp::cpp::extract_entities_windows_cpp": {
    "file": "tests\\analysis_cpp.rs",
    "module_path": "analysis_cpp::cpp",
    "test_function": "extract_entities_windows_cpp",
    "recorded_value": [
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Entity": {
                "description": {
                  "lines": [
                    "#TESTCLASS(c1)",
                    "#SET_UP_CONTEXT:",
                    "this->class_inst = new c1();",
                    "#TEAR_DOWN_CONTEXT:",
                    "delete this->class_inst;",
                    "this->class_inst = nullptr;",
                    "#CLASS_CONTEXT:",
                    "c1 * class_inst;"
                  ]
                },
                "entities": [
                  {
                    "Function": {
                      "arguments": [
                        {
                          "atype": "unsigned int",
                          "name": "no1",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no2",
                          "value": null
                        }
                      ],
                      "description": {
                        "lines": [
                          "#TESTCASE(c1::AddTwoNumbers)",
                          "unsigned int no1 = 5;",
                          "#EQ[this->class_inst->TL_FCT(no1:no1,no2:10)=>15]",
                          "#LT[this->class_inst->TL_FCT(no1:no1,no2:10)=>30]"
                        ]
                      },
                      "name": "add_two_numbers",
                      "return_type": "unsigned int"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c1",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c1",
                      "return_type": "void"
                    }
                  }
                ],
                "name": "c1"
              }
            },
            {
              "Entity": {
                "description": {
                  "lines": [
                    "#TESTCLASS(c2)",
                    "#SET_UP_CONTEXT:",
                    "this->class_inst = new c2();",
                    "#TEAR_DOWN_CONTEXT:",
                    "delete this->class_inst;",
                    "this->class_inst = nullptr;",
                    "#CLASS_CONTEXT:",
                    "c2 * class_inst;"
                  ]
                },
                "entities": [
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c2",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c2",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [
                        {
                          "atype": "unsigned int",
                          "name": "no1",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no2",
                          "value": null
                        },
                        {
                          "atype": "unsigned int",
                          "name": "no3",
                          "value": null
                        }
                      ],
                      "description": {
                        "lines": [
                          "#TESTCASE(c2::AddThreeNumbers)",
                          "unsigned int no1 = 5;",
                          "unsigned int no2 = 10;",
                          "#EQ[this->class_inst->TL_FCT(no1:no1,no2:no2,no3:5)=>20]"
                        ]
                      },
                      "name": "add_three_numbers",
                      "return_type": "unsigned int"
                    }
                  }
                ],
                "name": "c2"
              }
            }
          ],
          "name": "ns1"
        }
      },
      {
        "Entity": {
          "description": null,
          "entities": [
            {
              "Entity": {
                "description": null,
                "entities": [
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "c3",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "~c3",
                      "return_type": "void"
                    }
                  },
                  {
                    "Function": {
                      "arguments": [],
                      "description": null,
                      "name": "return5",
                      "return_type": "unsigned int"
                    }
                  }
                ],
                "name": "c3"
              }
            }
          ],
          "name": "ns2"
        }
      }
    ]
  }
}
